Ladle breaking TypeScript when `moduleResolution: "NodeNext"`

**Describe the bug**

Ladle breaks Typescript type-checking (e.g. `tsc --noEmit`) when Typescript has the config `moduleResolution: "NodeNext"`, even when `skipLibCheck: true`

**Reproduction**

https://stackblitz.com/edit/ladle-zq9ane?file=package.json

and run

```sh
npm run typecheck
```

The specific error that causes issues looks like the following, and note that it's looking in `node_modules`:

```cli
node_modules/@ladle/react/lib/app/exports.ts:2:33 - error TS2835: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean './src/context.js'?

2 import { useLadleContext } from "./src/context";
```

**Environment**

- OS: MacOS
- Browser: n/a
- Version: n/a

**More Context** 

From [this TS issue](https://github.com/microsoft/TypeScript/issues/41883) (that apparently is "working as intended" for now), it seems that the error comes from the fact that Ladle's `package.json#types` field [points to a `.ts` file](https://github.com/tajo/ladle/blob/4995ec02a78e9dc554f46b681c14cf6987e1237e/packages/ladle/package.json#L8) instead of a `.d.ts` file.

--- a/.changeset/tame-falcons-learn.md
+++ b/.changeset/tame-falcons-learn.md
@@ -1,0 +1,5 @@
+---
+"@ladle/react": patch
+---
+
+Update `package.json`'s "types" and "exports" so that they point to the type definition file, but only when publishing to NPM.

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -48,6 +48,6 @@ jobs:
           NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
 
       - name: Publishing next version
-        run: cd packages/ladle && ./publish-next.js
+        run: cd packages/ladle && pnpm types && ./publish-next.js
         env:
           NPM_TOKEN: ${{ secrets.NPM_TOKEN }}

--- a/packages/ladle/package.json
+++ b/packages/ladle/package.json
@@ -27,7 +27,8 @@
     "cli": "node ./lib/cli/cli.js",
     "clean": "rimraf dist && rimraf .ladle && rimraf build && rimraf *.tsbuildinfo",
     "serve": "node ./lib/cli/cli.js serve",
-    "test": "cross-env IMPORT_ROOT=\"./\" vitest"
+    "test": "cross-env IMPORT_ROOT=\"./\" vitest",
+    "types": "tsc --project tsconfig.typesoutput.json"
   },
   "dependencies": {
     "@babel/code-frame": "^7.18.6",

--- a/packages/ladle/publish-next.js
+++ b/packages/ladle/publish-next.js
@@ -2,6 +2,10 @@
 
 import { execSync } from "child_process";
 import fs from "fs";
+import {
+  preparePackageJsonForPublish,
+  revertPackageJson,
+} from "./scripts/package-types-helpers.js";
 
 const shortHash = execSync("git rev-parse --short HEAD").toString().trim();
 const version = `0.0.0-next-${shortHash}`;
@@ -13,6 +17,7 @@ const oldVersion = pkgJson.version;
 const oldContextVersion = pkgJson.dependencies["@ladle/react-context"];
 pkgJson.version = version;
 pkgJson.dependencies["@ladle/react-context"] = oldContextVersion.split(":")[1];
+preparePackageJsonForPublish(pkgJson);
 fs.writeFileSync("./package.json", JSON.stringify(pkgJson, null, 2));
 
 try {
@@ -24,4 +29,5 @@ try {
 
 pkgJson.version = oldVersion;
 pkgJson.dependencies["@ladle/react-context"] = oldContextVersion;
+revertPackageJson(pkgJson);
 fs.writeFileSync("./package.json", JSON.stringify(pkgJson, null, 2));

--- a/packages/ladle/scripts/package-types-helpers.js
+++ b/packages/ladle/scripts/package-types-helpers.js
@@ -1,0 +1,35 @@
+#!/usr/bin/env node
+
+let oldTypes = null;
+let oldExports = null;
+
+export function preparePackageJsonForPublish(packageJson) {
+  oldTypes = packageJson.types;
+  packageJson.types = "./lib/app/exports.d.ts";
+
+  oldExports = JSON.parse(JSON.stringify(packageJson.exports));
+  packageJson.exports["."] = {
+    types: "./lib/app/exports.d.ts",
+    default: "./lib/app/exports.ts",
+  };
+
+  return packageJson;
+}
+
+export function revertPackageJson(packageJson) {
+  if (!oldTypes) {
+    console.warn(`'oldTypes' is not defined, so we are unable to revert it`);
+  }
+
+  if (!oldExports) {
+    console.warn(`'oldExports' is not defined, so we are unable to revert it`);
+  }
+
+  packageJson.types = oldTypes ?? packageJson.types;
+  oldTypes = null;
+
+  packageJson.exports = oldExports ?? packageJson.exports;
+  oldExports = null;
+
+  return packageJson;
+}

--- a/packages/ladle/scripts/revert-package-types.js
+++ b/packages/ladle/scripts/revert-package-types.js
@@ -1,0 +1,12 @@
+import fs from "fs";
+
+const pkgJson = JSON.parse(
+  fs.readFileSync("./packages/ladle/backup-package.json"),
+);
+// write updates to package.json
+fs.writeFileSync(
+  "./packages/ladle/package.json",
+  JSON.stringify(pkgJson, null, 2),
+);
+// remove backup file
+fs.rmSync("./packages/ladle/backup-package.json");

--- a/packages/ladle/scripts/update-package-types.js
+++ b/packages/ladle/scripts/update-package-types.js
@@ -1,0 +1,16 @@
+import { preparePackageJsonForPublish } from "./package-types-helpers.js";
+import fs from "fs";
+
+const pkgJson = JSON.parse(fs.readFileSync("./packages/ladle/package.json"));
+// write out old package.json to a temp file that won't be published
+fs.writeFileSync(
+  "./packages/ladle/backup-package.json",
+  JSON.stringify(pkgJson, null, 2),
+);
+// update existing package.json
+preparePackageJsonForPublish(pkgJson);
+// write updates to package.json
+fs.writeFileSync(
+  "./packages/ladle/package.json",
+  JSON.stringify(pkgJson, null, 2),
+);

--- a/packages/ladle/tsconfig.typesoutput.json
+++ b/packages/ladle/tsconfig.typesoutput.json
@@ -1,0 +1,9 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "noEmit": false,
+    "declaration": true,
+    "emitDeclarationOnly": true
+  },
+  "include": ["lib/**/*", "src/**/*"]
+}

--- a/release.sh
+++ b/release.sh
@@ -7,5 +7,14 @@ echo "Running release"
 echo "Build @ladle/react"
 turbo run build --filter=@ladle/react
 
+echo "Update package.json"
+node ./packages/ladle/scripts/update-package-types.js
+
+echo "Generate Types":
+pnpm --filter @ladle/react types
+
 echo "Changeset publish"
 changeset publish
+
+echo "Revert package.json"
+node ./packages/ladle/scripts/revert-package-types.js

